File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def fun(self, mid, d, w):
        n=len(w)
        total=0
        day=1
        isvalid=True
        for i in range(n):
            if w[i]>mid:
                isvalid=False
            if total+w[i]<=mid:
                total+=w[i]
            else:
                day+=1
                total=w[i]
        if(not isvalid):
            return False
        else:
            return (day<=d)
        
        
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        l=1
        r=500*50000
        while(l<r):
            mid=(l+r)//2
            if(self.fun(mid,days,weights)):
                r=mid
            else:
                l=mid+1
        return l
            
        

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool fun(int mid,int d,vector<int>& w){
        int n=w.size();
        int day=1;
        int total=0;
        bool isvalid=true;
        for(int i=0;i<n;i++){
            if(w[i]>mid){
                isvalid=false;
            }
            if(total+w[i]<=mid){
                total+=w[i];
            }
            else{
                day++;
                total=w[i];
            }
        }
        if(!isvalid){
            return false;
        }
        else{
            return (day<=d);
        }
    }
    
    int shipWithinDays(vector<int>& weights, int days) {
        int l=1;
        int h=500*50000;
        int mid;
        while(l<h){
            mid=(l+h)/2;
            if(fun(mid,days,weights)){
                h=mid;
            }
            else{
                l=mid+1;
            }
        }
        return l;
    }
};